{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"text-align: right\">INFO 6105 Data Science Eng Methods and Tools, Lecture 9 Day 2</div>\n",
    "<div style=\"text-align: right\">Dino Konstantopoulos, 5 November 2020</div>\n",
    "\n",
    "# Why eugenics don't work\n",
    "\n",
    "Last class, where we explored the dark hsitory of statistics, I told you that diversity is more *powerful* than eugenics.\n",
    "\n",
    "What do I mean by powerful?\n",
    "\n",
    "Galton's claim that we can achieve a *better* society by selective breeding amongst humans included metrics of *intelligence, strength*, and *beauty*. \n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/ramanujan-ferrigno-zhiyi.png\" width=900 />\n",
    "</center>\n",
    "\n",
    "These characteristics are called [phenotypes](https://en.wikipedia.org/wiki/Phenotype) of a society, whereas the mechanism responsible for expressing these characteristics are each individual's [genotype](https://en.wikipedia.org/wiki/Genotype).\n",
    "\n",
    "Is Galton's statement truthful?\n",
    "\n",
    "We would like to see if we pick a population and a phenotype, is selectively breeding the population in order to achieve the best possible phenotype *the best possible way to achieve this*? In other words, is attempting to optimize the genotype the *fastest, most effective* way to the answer?\n",
    "\n",
    "Or will allowing the population to express all the diversity of its genotype by allowing the population to *freely* pick their own individuals to mate, in other words allowing the genotype to freely explore *all* possible configurations, achieve *better* and *quicker* results?\n",
    "\n",
    "So let's pick a phenotype we can express in *code*: Let's select an **integer** as the pinnacle of beauty/strength/intelligence. For example, I think the number `23` is incredibly *sexy*, don't you? \n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/23.jpg\" width=400 />\n",
    "</center>\n",
    "\n",
    "How do we reach that number? Well, with **addition** and **multiplication** of course! So let's breed a population that can learn to add and multiply in order to reach that incredibly sexy integer (or any other integer *you* find sexier)!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A Chromosome\n",
    "\n",
    "A chromosome is a binary string a little bit like this:\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/binary-chromosome.png\" width=500 />\n",
    "</center>\n",
    "\n",
    "Humans have 30,000 genes in their [genome](https://www.genome.gov/human-genome-project/Completion-FAQ). Each gene is made of two twisting paired strands. Each strand is made of four chemical units, called nucleotide bases. The bases are adenine (`A`), thymine (`T`), guanine (`G`) and cytosine (`C`). Bases on opposite strands pair specifically: An `A` always pairs with a `T`, and a `C` always with a `G`. \n",
    "\n",
    "So, let's assume our chemical units are 4 sequences of `0`s and `1`s :-) \n",
    "\n",
    ">**Question**: How many genes in the chromosome example above?\n",
    "\n",
    "At the beginning of a run of our genetic algorithm, a large population of random chromosomes will be created.\n",
    "\n",
    "Each chromosome, when decoded, will represent a different **solution** as to how to multiply and add in order to achieve that *sexy number*. That way, we have ourselves a *genotype* and a *phenotype*!\n",
    "\n",
    "To test each chromosome to see how good it is at solving the problem at hand, we assign it a **fitness score**. That is how we evaluate how *sexy* the chromosome is: Near or far-away from that pinnacle of sexiness, the number 23!\n",
    "\n",
    "That fitness score, of course, is a measure of how good that chromosome is at solving the problem to hand: Getting to our sexy number.\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/oh-no.jpg\" width=300 />\n",
    "</center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mating (reproduction)\n",
    "\n",
    "Reproduction will consist of chromosome **crossover**, performed by two genes, picking a point along the length of their chromosomes, and swapping genes after that point.\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/ga-reproduction.png\" width=400 />\n",
    "</center>\n",
    "\n",
    "Fred and Wilma give us bammbamm and pebbles :-)\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/flintstones-family.jpg\" width=400 />\n",
    "</center>\n",
    "\n",
    "Humans (as well as prehistoric humans) reproduce in... a bit more complicated way, of course ;-)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation\n",
    "\n",
    "Our chromosomes, just like humans (and Lou Ferrigno, a.k.a The [Hulk](https://en.wikipedia.org/wiki/Hulk)), sometimes *mutate*.\n",
    "\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/ga-mutate.png\" width=300 />\n",
    "</center>\n",
    "\n",
    "The **mutation rate** is the probability that a bit within a chromosome will be flipped (`0` becomes `1`, `1` becomes `0`). This is usually a very low value for binary encoded genes, say 0.001."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Encoding\n",
    "\n",
    "Four bits (a “short”) are required to represent the range of characters we'll use to add and multiply integers:\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/ga-encoding.png\" width=130 />\n",
    "</center>\n",
    "\n",
    "The possible genes `1110` and `1111` will remain unused and will be ignored by our algorithm if encountered.\n",
    "\n",
    "We will also ignore any non-sensical operations. For example, we expect a number to be followed by a math operation. If it doesn't, then the next gene (the 4 bits) will simply be *ignored* and we will move on to the next."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A sexy chromosome\n",
    "\n",
    "Here's a solution for getting to that pinnacle of sexiness: The number $23$: $6+5*4/2+1$:\n",
    "* $11 * 4 = 44$\n",
    "* $44/2 = 22$\n",
    "* $22 + 1 = 23$\n",
    "\n",
    "That solution would be represented by nine genes like so:\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/23-solution-1.png\" width=500 />\n",
    "</center>\n",
    "\n",
    ">**Note**: Arithmetic is read ***left to right*** (*not* the usual arithmetic associative rules)\n",
    "\n",
    "And thus a perfect sexy chromosome is:\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/23-solution-2.png\" width=500 />\n",
    "</center>\n",
    "\n",
    "But that's not the only one!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fitness function\n",
    "\n",
    "How do we grade our chromosomes for sexiness? Let's consider what happens when you're not the sexiest possible individual in the world...\n",
    "\n",
    "Let's move from Chicago to the Boston scene where `23` ain't that sexy anymore, and consider a *local* phenomenon of beauty: the number `42`:\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/al-horford-42.jpg\" width=500 />\n",
    "</center>\n",
    "\n",
    "Let's assign $\\infty$ fitness to chromosomes expressing our now-sexiest number 42, and the following fitness score to all others:\n",
    "\n",
    "$$\\text{score} = 1/(42 - \\text{bits-sum})$$\n",
    "\n",
    "So, our previously-beautiful chromosome 23...\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/23-solution-3.png\" width=500 />\n",
    "</center>\n",
    "\n",
    "..has a fitness score of $1/(42-23)$ or $1/19$.\n",
    "\n",
    "But, since infinity creates a slight problem with code, let's remormalize $1/0$ to $1/1$ when we encounter a $0$ in the denominator. \n",
    "\n",
    "So, the closer we get to the sexiest solution, the closer we get to 1, the sexier we are!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Roulette wheel selection\n",
    "\n",
    "Now we have to pick realistic strategies for 1) how a *free* society reproduces, 2) one where reproduction is dictated by forcing *sexiest* chromosomes to reproduce (eugenics), and 3) one where society is *stratified* into *castes* and forced to reproduce within a caste system (allowing for spurious exceptions).\n",
    "\n",
    "How individuals pick each other to mate in a free society picks is not a *new* algorithm. In fact, it is a well-known algorithm called [fitness-proportionate-selection](https://en.wikipedia.org/wiki/Fitness_proportionate_selection), also known as *roulette wheel selection*.\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/roulette-wheel-selection.png\" width=500 />\n",
    "</center>\n",
    "\n",
    "The algorithm recognizes that very sexy chromosomes will be attracted to each other, but leaves open the possibility that a super sexy chromosome will be attracted to a not-so-sexy chromosome. In other words, your professor still has a chance to marry [Rihana](https://en.wikipedia.org/wiki/Rihanna). Not a big chance, in other words, a low probability, but still a possibility!\n",
    "\n",
    "<br />\n",
    "<center>\n",
    "    <img src=\"ipynb.images/rihanna.jpg\" width=500 />\n",
    "</center>\n",
    "\n",
    "Ready? Let's code!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our hyperparameters, just a *swag* for now:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "CROSSOVER_RATE = 0.7\n",
    "MUTATION_RATE = 0.001\n",
    "POP_SIZE = 100 #must be an even number\n",
    "CHROMO_LENGTH = 300\n",
    "GENE_LENGTH = 4\n",
    "MAX_ALLOWABLE_GENERATIONS = 400"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our chromosome has bits and a fitness score:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "class chromo_typ:\n",
    "    #def __init__():\n",
    "    #    bits = ''\n",
    "    #    fitness = 0.0\n",
    "    def __init__(self, bts, ftns):\n",
    "        self.bits = bts\n",
    "        self.fitness = ftns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is how we build the chromosome bits, using a random number generator:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function returns a string of random 1s and 0s of desired length\n",
    "import random\n",
    "\n",
    "def GetRandomBits(length):\n",
    "    bits = []\n",
    "    for i in range(length):\n",
    "        x = random.random()\n",
    "        if x > 0.5:\n",
    "            bits.append(\"1\")\n",
    "        else:\n",
    "            bits.append(\"0\")\n",
    "    return ''.join(bits)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's an example chromosome consisting of 100 bits:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1000010001100010110101001111010011100100110100011101111010001000000011100110100001101011110011100011'"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example = GetRandomBits(100)\n",
    "example"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is how two chromosomes have sex and produce child chromosomes. Our chromosomes are very horny, but they only have sex 70% of the time ;-)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dependent on the CROSSOVER_RATE this function selects a random point along the \n",
    "# length of the chromosomes and swaps all the bits after that point.\n",
    "def Crossover(offspring1, offspring2):\n",
    "    newoffspring1 = ''\n",
    "    newoffspring2 = ''    \n",
    "    d = random.random()\n",
    "\n",
    "    # dependent on the crossover rate\n",
    "    if d < CROSSOVER_RATE:\n",
    "        #print(\"sex!\")\n",
    "        # sex! Create a random crossover point\n",
    "        crossover = int(d * CHROMO_LENGTH)\n",
    "\n",
    "        t1 = offspring1[:crossover] + offspring2[crossover:]\n",
    "        t2 = offspring2[:crossover] + offspring1[crossover:]\n",
    "\n",
    "        newoffspring1 = t1\n",
    "        newoffspring2 = t2\n",
    "        \n",
    "    else:\n",
    "        # read a book instead\n",
    "        newoffspring1 = offspring1\n",
    "        newoffspring2 = offspring2\n",
    "        \n",
    "    return newoffspring1, newoffspring2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Compare4Crossover(fred, offspring):\n",
    "    s = ''\n",
    "    for i in range(len(fred)):\n",
    "        if fred[i] == offspring[i]:\n",
    "            s += ' '\n",
    "        else:\n",
    "            s += '|'\n",
    "            break\n",
    "            \n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0010101001010100110000010111101111011101010010010010110000000000000000000001111101101100010010000110\n",
      "0000010011111110000101011111001100000001100000010101110000000001011010111010110000010011001100100111\n",
      "0010101001010100110000010111101111011101010010010010110000000000000000000010110000010011001100100111\n",
      "                                                                          |\n",
      "0000010011111110000101011111001100000001100000010101110000000001011010111001111101101100010010000110\n",
      "  |\n"
     ]
    }
   ],
   "source": [
    "fred = GetRandomBits(100)\n",
    "wilma = GetRandomBits(100)\n",
    "bammbamm, pebbles = Crossover(fred, wilma)\n",
    "print(fred)\n",
    "print(wilma)\n",
    "print(bammbamm)\n",
    "print(Compare4Crossover(fred, bammbamm))\n",
    "print(pebbles)\n",
    "print(Compare4Crossover(fred, pebbles))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sometimes, gamma ray radiation from the sun produces genetic mutations. You know, like the ones Darwin predicted for species:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mutates a chromosome's bits dependent on the MUTATION_RATE\n",
    "def Mutate(bits):\n",
    "    s = ''\n",
    "    for i in range(len(bits)):\n",
    "        d = random.random()\n",
    "        if d < MUTATION_RATE:\n",
    "            if bits[i] == '1':\n",
    "                s += '0'\n",
    "                # bits[i] = '0'\n",
    "            else:\n",
    "                s += '1'\n",
    "                # bits[i] = '1'\n",
    "        else:\n",
    "            s += bits[i]\n",
    "            \n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Compare4Mutation(bits1, bits2):\n",
    "    s = ''\n",
    "    for i in range(len(bits1)):\n",
    "        if bits1[i] == bits2[i]:\n",
    "            s += ' '\n",
    "        else:\n",
    "            s += '|'\n",
    "            \n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000010001100010110101001111010011100100110100011101111010001000000011100110100001101011110011100011\n",
      "1000010001100010110101001111010011100100110100011101111010001000000011100110100001101011110011100011\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'                                                                                                    '"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(example)\n",
    "newexample = Mutate(example)\n",
    "print(newexample)\n",
    "Compare4Mutation(example, newexample)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And here is how chromosomes are selected to either have sex or to spontaneously mutate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selects a chromosome from the population via roulette wheel selection\n",
    "# population is a list of chromo_typ\n",
    "def Roulette(total_fitness, population):\n",
    "    # generate a random number between 0 & total fitness count, based on the input seed\n",
    "    d = random.random() \n",
    "    slice = d * total_fitness\n",
    "\n",
    "    # go through the chromosomes adding up the fitness so far\n",
    "    fitnessSoFar = 0.0\n",
    "\n",
    "    for i in range(POP_SIZE):\n",
    "        fitnessSoFar += population[i].fitness\n",
    "\n",
    "        # if the fitness so far > random number return the chromo at this point\n",
    "        if fitnessSoFar >= slice:\n",
    "            return population[i].bits\n",
    "    return '(not fit enough)'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Transcription layer\n",
    "\n",
    "Here is our transcription layer. It's the equivalent of RNA to DNA transcription in living species on our planet:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "def BinToDec(bits):\n",
    "    val = 0\n",
    "    value_to_add = 1\n",
    "\n",
    "    for i in range(len(bits), 0, -1):\n",
    "        if bits[i - 1] == '1':\n",
    "            val += value_to_add\n",
    "        value_to_add *= 2\n",
    "\n",
    "    return val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Given a chromosome, this function will step through the genes one at a time and insert \n",
    "#  the decimal values of each gene (which follow the number -> operator -> number -> operator rule)\n",
    "#  into a buffer. Returns the number of elements in the buffer, and the buffer as an out arg.\n",
    "#  In other words, this is our DNA to RNA layer!\n",
    "def ParseBits(bits, interpret = False):\n",
    "    # buffer\n",
    "    buffer = []\n",
    "    # counter for buffer position\n",
    "    cBuff = 0\n",
    "\n",
    "    #  step through bits a gene at a time until end and store decimal values\n",
    "    #  of valid operators and numbers. Don't forget we are looking for  \n",
    "    #  number - operator - number - operator and so on... \n",
    "    #  We ignore unused genes 1111 and 1110\n",
    "\n",
    "    #  flag to determine if we are looking for an operator or a number\n",
    "    bOperator = False\n",
    "\n",
    "    #  storage for decimal value of currently tested gene\n",
    "    this_gene = 0\n",
    "\n",
    "    for i in range(0, CHROMO_LENGTH, GENE_LENGTH):\n",
    "        # convert the current gene to decimal\n",
    "        this_gene = BinToDec(bits[i: i + GENE_LENGTH])\n",
    "\n",
    "        # find a gene which represents an operator\n",
    "        if bOperator:\n",
    "            if this_gene < 10 or this_gene > 13:\n",
    "                continue\n",
    "            else:\n",
    "                if not interpret:\n",
    "                    buffer.append(this_gene)\n",
    "                else:\n",
    "                    buffer.append('+' if this_gene == 10 else '-' if this_gene == 11 else '*' if this_gene == 12 else '/')\n",
    "                cBuff += 1\n",
    "                bOperator = False\n",
    "                # buffer.Add(this_gene)\n",
    "\n",
    "        # find a gene which represents a number\n",
    "        else:\n",
    "            if this_gene > 9:\n",
    "                continue\n",
    "            else:\n",
    "                buffer.append(this_gene)\n",
    "                cBuff += 1\n",
    "                bOperator = True\n",
    "\n",
    "\n",
    "    #     now we have to run through buffer to see if a possible divide by zero\n",
    "    #     is included and delete it. (ie a '/' followed by a '0'). We take an easy\n",
    "    #     way out here and just change the '/' to a '+'. This will not effect the \n",
    "    #     evolution of the solution\n",
    "    for i in range(cBuff - 1):\n",
    "        if not interpret:\n",
    "            if buffer[i] == 13 and buffer[i + 1] == 0:\n",
    "                buffer[i] = 10\n",
    "        else:\n",
    "            if type(buffer[i]) is str:\n",
    "                if buffer[i] == '/' and buffer[i + 1] == 0:\n",
    "                    buffer[i] = '+'\n",
    "\n",
    "    #  debugging\n",
    "    #chromosome = DisplayChromo(buffer)\n",
    "\n",
    "    #return cBuff\n",
    "    return buffer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type('*') is str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100001110110010101011101100110110000111110011110100101101011010001011101011 [8, '/', 9, '-', 0, '-', 4, '/', 3]\n"
     ]
    }
   ],
   "source": [
    "fred = GetRandomBits(75)\n",
    "print(fred, ParseBits(fred, interpret = True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  given a chromosome (of bits) this function will calculate its phenotype (representation)\n",
    "# 1st implementation\n",
    "def Phenotype_old(bits):\n",
    "    buffer = ParseBits(bits, CHROMO_LENGTH // GENE_LENGTH)\n",
    "\n",
    "    #  ok, we have a buffer filled with valid values of: \n",
    "    #  operator - number - operator - number..\n",
    "    #  now we calculate what this represents\n",
    "    \n",
    "    result = 0\n",
    "    for i in range(0, len(buffer) - 1, 2):\n",
    "        buffer_i = buffer[i]\n",
    "        if buffer_i == 10:\n",
    "            result += buffer[i + 1]\n",
    "\n",
    "        if buffer_i == 11:\n",
    "            result -= buffer[i + 1]\n",
    "\n",
    "        if buffer_i == 12:\n",
    "            result *= buffer[i + 1]\n",
    "\n",
    "        if buffer_i == 13:\n",
    "            result //= buffer[i + 1]\n",
    "                \n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  given a chromosome (of bits) this function will calculate its phenotype (representation)\n",
    "def Phenotype(bits):\n",
    "    # List<int> buffer2 = new List<int>();\n",
    "    # num_elements = len(buffer)\n",
    "    buffer = ParseBits(bits)\n",
    "\n",
    "    #  ok, we have a buffer filled with valid values of: \n",
    "    #  number - operator - number - operator..\n",
    "    #  now we calculate what this represents\n",
    "    \n",
    "    result = 0\n",
    "    if len(buffer) < 3:\n",
    "        return result\n",
    "    \n",
    "    buffer_i = buffer[1]\n",
    "    \n",
    "    # first operation\n",
    "    if buffer_i == 10:\n",
    "        result = buffer[0] + buffer[2] \n",
    "\n",
    "    if buffer_i == 11:\n",
    "        result = buffer[0] - buffer[2]\n",
    "\n",
    "    if buffer_i == 12:\n",
    "        result = buffer[0] * buffer[2]\n",
    "\n",
    "    if buffer_i == 13:\n",
    "        result = buffer[0] // buffer[2]\n",
    "    \n",
    "    # subsequent operations\n",
    "    for i in range(3, len(buffer) - 1, 2):\n",
    "        buffer_i = buffer[i]\n",
    "        \n",
    "        if buffer_i == 10:\n",
    "            result += buffer[i + 1] \n",
    "\n",
    "        if buffer_i == 11:\n",
    "            result -= buffer[i + 1]\n",
    "\n",
    "        if buffer_i == 12:\n",
    "            result *= buffer[i + 1]\n",
    "\n",
    "        if buffer_i == 13:\n",
    "            result //= buffer[i + 1]\n",
    "            \n",
    "        # debugging\n",
    "        #print(result)\n",
    "                \n",
    "    return abs(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('1101111010000111001000011001001010100011100000000100100001111010001100001101000100101000011011010101',\n",
       " [8, '+', 3, '+', 3, '/', 1, '/', 5],\n",
       " 2)"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example = GetRandomBits(100)\n",
    "parseexample = ParseBits(example, interpret = True)\n",
    "example, parseexample, Phenotype(example)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  given a phenotype (chromosome representation) and a target value, this function will assign and return a fitness score\n",
    "def AssignFitness(chromosome, target_value):\n",
    "    result = Phenotype(chromosome.bits)\n",
    "\n",
    "    #  Now we calculate the fitness. First check to see if a solution has been found\n",
    "    #  and assign an arbitarily high fitness score if this is so.\n",
    "\n",
    "    if result == target_value:\n",
    "        chromosome.fitness = 1.0\n",
    "    else:\n",
    "        chromosome.fitness = 1. / (abs(target_value - result) + 1.)\n",
    "        \n",
    "    return chromosome.fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "1. / (abs(42 - 42) + 1), 1. / (abs(41 - 42) + 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper functions\n",
    "\n",
    "For better debugging!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DisplayChromo_old(bits):\n",
    "    # parse the bit string\n",
    "    buffer = ParseBits(bits)\n",
    "\n",
    "    s = ''\n",
    "    for i in range(len(buffer)):\n",
    "        s += DisplayGeneSymbol(buffer[i])\n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DisplayChromo(bits):\n",
    "    return ' '.join([str(x) for x in ParseBits(bits, interpret = True)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DisplayGeneSymbol(val):\n",
    "    s = ''\n",
    "    if int(val) < 10:\n",
    "        s += val\n",
    "        \n",
    "    else:\n",
    "        if int(val) == 10:\n",
    "            s += '+'\n",
    "\n",
    "        if int(val) == 11:\n",
    "            s += '-'\n",
    "\n",
    "        if int(val) == 12:\n",
    "            s += '*'\n",
    "\n",
    "        if int(val) == 13:\n",
    "            s += '/'\n",
    "\n",
    "    s += ' '\n",
    "    return s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genetic algorithm driver\n",
    "\n",
    "We start with some *low* hyperparameter values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "# debugging\n",
    "CHROMO_LENGTH = 40\n",
    "POP_SIZE = 10\n",
    "MAX_ALLOWABLE_GENERATIONS = 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We transcribe and evaluate the fitness of a population of 10 chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1100111101100010101001111111110001001100\n",
      "0001001000010101001111010101101011000001\n",
      "1111010011100011100111101111010101001111\n",
      "1001101001101111000001100110011010100100\n",
      "1011110100000101111010000111001110011000\n",
      "0011111111000010100000011000100001011010\n",
      "0000000100110010101100000010101100010110\n",
      "1010000111011010110010100001000000010111\n",
      "1010000110111110001101010011110010011111\n",
      "0111100101000011101010001111101101101000\n",
      "\n",
      "12-15-6-2-10-7-15-12-4-12-\n",
      "0\n",
      "1-2-1-5-3-13-5-10-12-1-\n",
      "0\n",
      "15-4-14-3-9-14-15-5-4-15-\n",
      "0\n",
      "9-10-6-15-0-6-6-6-10-4-\n",
      "0\n",
      "11-13-0-5-14-8-7-3-9-8-\n",
      "0\n",
      "3-15-12-2-8-1-8-8-5-10-\n",
      "0\n",
      "0-1-3-2-11-0-2-11-1-6-\n",
      "4\n",
      "10-1-13-10-12-10-1-0-1-7-\n",
      "4\n",
      "10-1-11-14-3-5-3-12-9-15-\n",
      "4\n",
      "7-9-4-3-10-8-15-11-6-8-\n",
      "0\n",
      "\n",
      "[6, 10, 7, 12, 4, 12]\n",
      "[1, 13, 5, 10, 1]\n",
      "[4]\n",
      "[9, 10, 6, 10, 4]\n",
      "[0]\n",
      "[3, 12, 2, 10]\n",
      "[0, 11, 0, 11, 1]\n",
      "[1, 13, 1]\n",
      "[1, 11, 3, 12, 9]\n",
      "[7, 10, 8, 11, 6]\n",
      "\n",
      "[6, '+', 7, '*', 4, '*'] 52 0.1\n",
      "[1, '/', 5, '+', 1] 1 0.024390243902439025\n",
      "[4] 0 0.023809523809523808\n",
      "[9, '+', 6, '+', 4] 19 0.043478260869565216\n",
      "[0] 0 0.023809523809523808\n",
      "[3, '*', 2, '+'] 6 0.027777777777777776\n",
      "[0, '-', 0, '-', 1] 1 0.024390243902439025\n",
      "[1, '/', 1] 1 0.024390243902439025\n",
      "[1, '-', 3, '*', 9] 18 0.041666666666666664\n",
      "[7, '+', 8, '-', 6] 9 0.030303030303030304\n",
      "\n",
      "0 0.1\n",
      "1 0.024390243902439025\n",
      "2 0.023809523809523808\n",
      "3 0.043478260869565216\n",
      "4 0.023809523809523808\n",
      "5 0.027777777777777776\n",
      "6 0.024390243902439025\n",
      "7 0.024390243902439025\n",
      "8 0.041666666666666664\n",
      "9 0.030303030303030304\n",
      "\n",
      "Selecting 20 fit individuals:\n",
      "(not fit enough)\n",
      "(not fit enough)\n",
      "0111100101000011101010001111101101101000\n",
      "0011111111000010100000011000100001011010\n",
      "(not fit enough)\n",
      "1011110100000101111010000111001110011000\n",
      "(not fit enough)\n",
      "0111100101000011101010001111101101101000\n",
      "(not fit enough)\n",
      "(not fit enough)\n",
      "1001101001101111000001100110011010100100\n",
      "1100111101100010101001111111110001001100\n",
      "1010000111011010110010100001000000010111\n",
      "(not fit enough)\n",
      "(not fit enough)\n",
      "(not fit enough)\n",
      "1001101001101111000001100110011010100100\n",
      "1100111101100010101001111111110001001100\n",
      "(not fit enough)\n",
      "(not fit enough)\n"
     ]
    }
   ],
   "source": [
    "for j in range(1):\n",
    "    #  storage for our population of chromosomes.\n",
    "    # size: [POP_SIZE]\n",
    "    #List<chromo_typ> Population = new List<chromo_typ>()\n",
    "    population = []\n",
    "\n",
    "    #  specify our sexy target number\n",
    "    sexyTarget= 42\n",
    "\n",
    "    #  first create a random population, all with zero fitness.\n",
    "    for i in range(POP_SIZE):\n",
    "        ct = chromo_typ(GetRandomBits(CHROMO_LENGTH), 0.0)\n",
    "        print(ct.bits)\n",
    "        population.append(ct) \n",
    "\n",
    "print()\n",
    "for i in range(POP_SIZE):\n",
    "    repr = ''\n",
    "    arr = []\n",
    "    for j in range(0, CHROMO_LENGTH, GENE_LENGTH):\n",
    "        # convert the current gene to decimal\n",
    "        quad = BinToDec(population[i].bits[j: j + GENE_LENGTH])\n",
    "        arr.append(str(quad))\n",
    "        repr += str(quad) + '-'\n",
    "    print(repr)\n",
    "    print(DisplayChromo(arr))\n",
    "    \n",
    "print()\n",
    "for i in range(POP_SIZE):\n",
    "    #arr = ParseBits(population[i].bits, CHROMO_LENGTH // GENE_LENGTH)\n",
    "    arr = ParseBits(population[i].bits)\n",
    "    print(arr)\n",
    "\n",
    "print()\n",
    "for i in range(POP_SIZE):\n",
    "    print(ParseBits(population[i].bits, interpret = True), Phenotype(population[i].bits), AssignFitness(population[i], 42))\n",
    "\n",
    "print()\n",
    "for i in range(POP_SIZE):\n",
    "    print(i, population[i].fitness)\n",
    "    \n",
    "print()\n",
    "print('Selecting 20 fit individuals:')\n",
    "for i in range(20):\n",
    "    print(Roulette(random.randint(1, POP_SIZE//10), population))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We're now ready for a generation loop:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#CROSSOVER_RATE = 0.7\n",
    "CROSSOVER_RATE = 0.6\n",
    "MUTATION_RATE = 0.001\n",
    "#POP_SIZE = 100 #must be an even number\n",
    "POP_SIZE = 10\n",
    "#CHROMO_LENGTH = 300\n",
    "CHROMO_LENGTH = 30\n",
    "GENE_LENGTH = 4\n",
    "#MAX_ALLOWABLE_GENERATIONS = 400\n",
    "MAX_ALLOWABLE_GENERATIONS = 400"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best chromosome detected in generation 0, has fitness 0.02857142857142857, phenotype 7, and genotype 1 * 7 -\n",
      "Best chromosome detected in generation 1, has fitness 0.02857142857142857, phenotype 7, and genotype 7 * 1\n",
      "Best chromosome detected in generation 2, has fitness 0.02857142857142857, phenotype 7, and genotype 7 * 1\n",
      "Best chromosome detected in generation 3, has fitness 0.027777777777777776, phenotype 6, and genotype 4 + 2\n",
      "Best chromosome detected in generation 4, has fitness 0.03571428571428571, phenotype 14, and genotype 4 * 4 + 2 - 4\n",
      "Best chromosome detected in generation 5, has fitness 0.05555555555555555, phenotype 24, and genotype 4 * 7 - 4\n",
      "Best chromosome detected in generation 6, has fitness 0.043478260869565216, phenotype 19, and genotype 7 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 7, has fitness 0.05263157894736842, phenotype 23, and genotype 7 * 3 + 2\n",
      "Best chromosome detected in generation 8, has fitness 0.02702702702702703, phenotype 5, and genotype 7 + 2 - 4\n",
      "Best chromosome detected in generation 9, has fitness 0.02702702702702703, phenotype 5, and genotype 7 + 2 - 4\n",
      "Best chromosome detected in generation 10, has fitness 0.03125, phenotype 10, and genotype 4 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 11, has fitness 0.03125, phenotype 10, and genotype 4 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 12, has fitness 0.03125, phenotype 10, and genotype 4 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 13, has fitness 0.03125, phenotype 10, and genotype 4 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 14, has fitness 0.02564102564102564, phenotype 3, and genotype 7 - 4\n",
      "Best chromosome detected in generation 15, has fitness 0.02564102564102564, phenotype 3, and genotype 7 - 4\n",
      "Best chromosome detected in generation 16, has fitness 0.02702702702702703, phenotype 5, and genotype 7 + 2 - 4\n",
      "Best chromosome detected in generation 17, has fitness 0.02702702702702703, phenotype 5, and genotype 7 + 2 - 4\n",
      "Best chromosome detected in generation 18, has fitness 0.029411764705882353, phenotype 8, and genotype 4 * 3 - 4\n",
      "Best chromosome detected in generation 19, has fitness 0.029411764705882353, phenotype 8, and genotype 4 * 3 - 4\n",
      "Best chromosome detected in generation 20, has fitness 0.029411764705882353, phenotype 8, and genotype 4 * 3 - 4\n",
      "Best chromosome detected in generation 21, has fitness 0.029411764705882353, phenotype 8, and genotype 4 * 3 - 4\n",
      "Best chromosome detected in generation 22, has fitness 0.029411764705882353, phenotype 8, and genotype 4 * 3 - 4\n",
      "Best chromosome detected in generation 23, has fitness 0.029411764705882353, phenotype 8, and genotype 4 * 3 - 4\n",
      "Best chromosome detected in generation 24, has fitness 0.04, phenotype 17, and genotype 7 * 3 - 4\n",
      "Best chromosome detected in generation 25, has fitness 0.04, phenotype 17, and genotype 7 * 3 - 4\n",
      "Best chromosome detected in generation 26, has fitness 0.04, phenotype 17, and genotype 7 * 3 - 4\n",
      "Best chromosome detected in generation 27, has fitness 0.04, phenotype 17, and genotype 7 * 3 - 4\n",
      "Best chromosome detected in generation 28, has fitness 0.04, phenotype 17, and genotype 7 * 3 - 4\n",
      "Best chromosome detected in generation 29, has fitness 0.045454545454545456, phenotype 20, and genotype 6 * 3 + 6 - 4\n",
      "Best chromosome detected in generation 30, has fitness 0.043478260869565216, phenotype 19, and genotype 7 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 31, has fitness 0.043478260869565216, phenotype 19, and genotype 7 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 32, has fitness 0.043478260869565216, phenotype 19, and genotype 7 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 33, has fitness 0.043478260869565216, phenotype 19, and genotype 7 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 34, has fitness 0.043478260869565216, phenotype 19, and genotype 7 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 35, has fitness 0.043478260869565216, phenotype 19, and genotype 7 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 36, has fitness 0.043478260869565216, phenotype 19, and genotype 7 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 37, has fitness 0.03125, phenotype 10, and genotype 4 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 38, has fitness 0.03125, phenotype 10, and genotype 4 * 3 + 2 - 4\n",
      "Best chromosome detected in generation 39, has fitness 0.0625, phenotype 26, and genotype 4 * 7 + 2 - 4\n",
      "Best chromosome detected in generation 40, has fitness 0.0625, phenotype 26, and genotype 4 * 7 + 2 - 4\n",
      "Best chromosome detected in generation 41, has fitness 0.0625, phenotype 26, and genotype 4 * 7 + 2 - 4\n",
      "Best chromosome detected in generation 42, has fitness 0.1111111111111111, phenotype 33, and genotype 5 * 7 + 2 - 4\n",
      "Best chromosome detected in generation 43, has fitness 0.1111111111111111, phenotype 33, and genotype 5 * 7 + 2 - 4\n",
      "Best chromosome detected in generation 44, has fitness 0.1111111111111111, phenotype 33, and genotype 5 * 7 + 2 - 4\n",
      "Best chromosome detected in generation 45, has fitness 0.1111111111111111, phenotype 33, and genotype 5 * 7 + 2 - 4\n",
      "Best chromosome detected in generation 46, has fitness 0.1111111111111111, phenotype 33, and genotype 5 * 7 + 2 - 4\n",
      "Best chromosome detected in generation 47, has fitness 0.125, phenotype 34, and genotype 5 * 7 + 3 - 4\n",
      "Best chromosome detected in generation 48, has fitness 0.125, phenotype 34, and genotype 5 * 7 + 3 - 4\n",
      "Best chromosome detected in generation 49, has fitness 0.125, phenotype 34, and genotype 5 * 7 + 3 - 4\n",
      "Best chromosome detected in generation 50, has fitness 0.125, phenotype 34, and genotype 5 * 7 + 3 - 4\n",
      "Best chromosome detected in generation 51, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 52, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 53, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 54, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 55, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 56, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 57, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 58, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 59, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 60, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 61, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 62, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 63, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 64, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 65, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 66, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 67, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 68, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 69, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 70, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 71, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 72, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 73, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 74, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 75, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 76, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 77, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 78, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 79, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 80, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 81, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 82, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 83, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 84, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 85, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 86, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 87, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 88, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 89, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 90, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 91, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 92, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 93, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 94, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 95, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 96, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 97, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 98, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 99, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "Best chromosome detected in generation 100, has fitness 1.0, phenotype 41, and genotype 5 * 7 + 2 + 4\n",
      "=========================\n",
      "No exact solutions found!\n",
      "Best chromosome found has fitness 1.0 , phenotype 41 , genotype 5 * 7 + 2 + 4 , and is the chromosome 0101111111000111101000101010101101000001\n",
      "=========================\n",
      "\n",
      "*** Next round:\n"
     ]
    }
   ],
   "source": [
    "# How much info to print: 0 or 1\n",
    "VERBOSITY = 0\n",
    "\n",
    "#  overall champ\n",
    "ultimatechamp = chromo_typ('', 0.0) \n",
    "\n",
    "# just loop endlessly \n",
    "#while true:\n",
    "for j in range(1):\n",
    "    #  storage for our population of chromosomes.\n",
    "    # size: [POP_SIZE]\n",
    "    #List<chromo_typ> Population = new List<chromo_typ>()\n",
    "    population = []\n",
    "\n",
    "    #  specify our sexy target number\n",
    "    sexyTarget= 42\n",
    "\n",
    "    #  first create a random population, all with zero fitness.\n",
    "    for i in range(POP_SIZE):\n",
    "        ct = chromo_typ(GetRandomBits(CHROMO_LENGTH), 0.0)\n",
    "        population.append(ct) \n",
    "\n",
    "    generationsRequiredToFindASolution = 0\n",
    "\n",
    "    # we will set this flag if a solution has been found\n",
    "    bFound = False\n",
    "\n",
    "    # run one experiment up to MAX_ALLOWABLE_GENERATIONS\n",
    "    lastchamp = chromo_typ('', 0.0)\n",
    "    while not bFound:\n",
    "        # this is used during roulette wheel sampling\n",
    "        totalFitness = 0.0\n",
    "\n",
    "        #  test and update the fitness of every chromosome in the population\n",
    "        for ct in population:\n",
    "            ct.fitness = AssignFitness(ct, sexyTarget)\n",
    "            totalFitness += ct.fitness\n",
    "\n",
    "        #  check to see if we have found any solutions\n",
    "        thechamp = chromo_typ('', 0.0)\n",
    "        for ct in population:\n",
    "            #if ct.fitness == 999.0:\n",
    "            if Phenotype(ct.bits) == sexyTarget:\n",
    "                ultimatechamp.fitness = ct.fitness\n",
    "                ultimatechamp.bits = ct.bits\n",
    "\n",
    "                print(\"=========================\")\n",
    "                print(\"Solution found in \" + str(generationsRequiredToFindASolution) + \" generations! Its genes are:\", \n",
    "                      DisplayChromo(ct.bits))\n",
    "                print(\"It has phenotype\", str(Phenotype(ct.bits)), \", fitness\", str(ct.fitness), \", and chromosome\", ct.bits)\n",
    "                print(\"=========================\")\n",
    "                bFound = True\n",
    "                break\n",
    "                \n",
    "            else:\n",
    "                # current experiment champ\n",
    "                if ct.fitness > thechamp.fitness:\n",
    "                    thechamp.fitness = ct.fitness\n",
    "                    thechamp.bits = ct.bits; \n",
    "                # all experiments champ\n",
    "                if ct.fitness > ultimatechamp.fitness:\n",
    "                    ultimatechamp.fitness = ct.fitness\n",
    "                    ultimatechamp.bits = ct.bits; \n",
    "        if not bFound:\n",
    "            if VERBOSITY == 1:\n",
    "                print(\"Best chromosome detected in generation \" + str(generationsRequiredToFindASolution) + \n",
    "                      \", has fitness \" + str(thechamp.fitness) + \", phenotype \" + str(Phenotype(thechamp.bits)) + \n",
    "                      \", and genotype \"  + DisplayChromo(thechamp.bits)) \n",
    "            elif VERBOSITY == 0:\n",
    "                if thechamp.fitness > lastchamp.fitness or generationsRequiredToFindASolution % (MAX_ALLOWABLE_GENERATIONS // 5) == 0:\n",
    "                    print(\"Best chromosome detected in generation \" + str(generationsRequiredToFindASolution) + \n",
    "                          \", has fitness \" + str(thechamp.fitness) + \", phenotype \" + str(Phenotype(thechamp.bits)) + \n",
    "                          \", and genotype \"  + DisplayChromo(thechamp.bits)) \n",
    "            lastchamp.fitness = thechamp.fitness\n",
    "            lastchamp.bits = thechamp.bits\n",
    "\n",
    "        # \n",
    "        #  We did not find our champion :-(\n",
    "        #  create a new population by selecting two parents at a time and creating offspring\n",
    "        #  by applying crossover and mutation. Do this until the desired number of offspring\n",
    "        #  have been created\n",
    "        # \n",
    "\n",
    "        # define some temporary storage for the new population we are about to create\n",
    "        # chromo_typ temp[POP_SIZE]\n",
    "        #List<chromo_typ> temp = new List<chromo_typ>()\n",
    "        temp = []\n",
    "\n",
    "        # loop until we have created POP_SIZE new chromosomes\n",
    "        cPop = 0\n",
    "        while cPop < POP_SIZE:\n",
    "            #  we are going to create the new population by grabbing members of the old \n",
    "            #  population two at a time via roulette wheel selection. Individuals are selected\n",
    "            #  for crossover by a combination of fitness and luck\n",
    "            #Random r = new Random(GenerationsRequiredToFindASolution)\n",
    "            fred = Roulette(totalFitness, population) #maybe this is not too correct: pick a fred at random instead?\n",
    "            wilma = Roulette(totalFitness, population)\n",
    "            if fred == '(not fit enough)' or wilma == '(not fit enough)':\n",
    "                continue\n",
    "\n",
    "            #  add crossover dependent on the crossover rate\n",
    "            bammbamm, pebbles = Crossover(fred, wilma)\n",
    "\n",
    "            #  mutate dependent on the mutation rate\n",
    "            bammbamm = Mutate(bammbamm)\n",
    "            pebbles = Mutate(pebbles)\n",
    "\n",
    "            # add these offspring to the new population. (assigning zero as their fitness scores)\n",
    "            # Note that if fred and wilma do create offspring, then fred and wilma essentially die off :-(\n",
    "            ct1 = chromo_typ(bammbamm, 0.0)\n",
    "            ct2 = chromo_typ(pebbles, 0.0)\n",
    "            temp.append(ct1); cPop += 1\n",
    "            temp.append(ct2); cPop += 1\n",
    "\n",
    "\n",
    "        # copy new population into main population array\n",
    "        population.clear()\n",
    "        population.extend(temp)\n",
    "        #for i in range(POP_SIZE):\n",
    "        #    population.append(temp[i])\n",
    "\n",
    "        generationsRequiredToFindASolution += 1\n",
    "\n",
    "        #  exit experiment if no solution found within the maximum allowable number\n",
    "        #  of generations\n",
    "        if generationsRequiredToFindASolution > MAX_ALLOWABLE_GENERATIONS:\n",
    "            print(\"-------------------------\")\n",
    "            print(\"No exact solutions found!\")\n",
    "            print(\"Best chromosome found has fitness\", str(round(ultimatechamp.fitness, 2)), \", phenotype\", \n",
    "                  str(Phenotype(ultimatechamp.bits)), \", genotype\", DisplayChromo(ultimatechamp.bits), \n",
    "                  \", and bits\", ultimatechamp.bits)\n",
    "            print(\"-------------------------\")\n",
    "            bFound = True\n",
    "    print(\"\")\n",
    "    print(\"*** Next round:\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks to me that we kinda get *stuck* in local minima. Maybe my hyperparameters are not so good?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework\n",
    "\n",
    "With the right hyperparameters, compare how a *free* society reproduces with one where reproduction is dictated by forcing *sexiest* chromosomes to reproduce (eugenics), and one where society is *stratified* into *castes* and forced to reproduce within a caste system (allowing for spurious exceptions).\n",
    "\n",
    "The objective of your hw is to compare eugenics, castes, and roulette wheel selection.\n",
    "\n",
    "Roulette wheel is how species freely evolve in a Darwinian way.\n",
    "\n",
    "Eugenic proponents, Nazi Germany amongst them, suggest that a superior race is created when we breed individuals with the best fitness, and we kill off individuals with poor fitness. Write the equivalent algorithm to Roulette Wheel selection for eugenics.\n",
    "\n",
    "Caste-based societies evolve with rules, not allowing members of inferior fitness to reproduce with member of superior fitness. Assume a certain number of castes and write the equivalent of Roulette wheel selection for caste-based societies. Allow for a few escapees from the \"Law\".\n",
    "\n",
    "Then, do data science. Run many experiments with different examples of sexiness, different hyperparameters, and compare all 3 societies. Which society achieves sexy target the fastest? What happens to the fitness of the rest of society? Discuss, and conclude."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
